### PR TITLE
Add docs for artifact proxy

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## What changes are proposed in this pull request?

Adding documentation for proxy artifact handling mode for the MLflow server. 

## How is this patch tested?

sphinx doc build in local

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Documentation added for using the Mlflow server in proxy artifact mode and in exclusive artifact handling mode.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:86]: This is an MLflow server configuration, not a client configuration.

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:187]: Nit: "File store" is a bit of an overloaded term in MLflow, generally referring to the `FileStore` implementation of the backend metadata store API. Can we use "object store" instead?

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:189]: The "localhost" box in the `scenario_5` and `scenario_6` images show lines connecting from "RestStore" to "Remote Host. "RestStore" refers to the REST-based implementation of the backend metadata storage API, not an artifact storage API implementation. We should instead add a second small box referring to `HttpArtifactRepository` (https://github.com/mlflow/mlflow/blob/3c434246f73b41835194f3e6aa1f9be8488bcd56/mlflow/store/artifact/http_artifact_repo.py#L11).

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:765]: Is this right? I think this should be `--artifacts-destination`.

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:1048]:
```suggestion
        --artifacts-destination hdfs://myhost:8887/mlprojects/models \
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:1053]: There are other methods used for logging artifacts, e.g. `mlflow.log_artifact()`. Perhaps we should update the phrasing to indicate that this isn't an exhaustive list of artifact-based APIs. This is also Python-specific.

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:836]: Nit: Instead of using future-tense "will" throughout the docs here, can we use present-tense phrasing (not an English expert, these are probably inaccurate tense names):

```suggestion
Starting a server with the ``--serve-artifacts`` option defined enables proxy access for artifacts.
```

One more small phrasing suggestion:

```suggestion
Starting a server with the ``--serve-artifacts`` option defined enables proxied access for artifacts.
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:838]:
```suggestion
The URI ``mlflow-artifacts:/`` is the default value for ``--default-artifact-root`` when `--serve-artifacts` is specified. This indicates to clients that artifacts can be accessed via HTTP requests to the MLflow Tracking Server.
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:843]: These first two sentences are redundant given the information above, right?

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:1034]:
```suggestion
Using the Tracking Server exclusively for proxied artifact access
```

File: docs/source/tracking.rst
Comment (dbczumar) [reply docs/source/tracking.rst:838]: Can we also enumerate and explain the use cases for some other proxied artifact `--default-artifact-root` values? E.g. `mlflow-artifacts://<tracking_server_host_name>/sub/path` or `https://<host_name>/sub/path`?

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:187]: fixed and added references to object stores (just in case)

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:86]: yikes yeah. fixed.

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:206]: Specifically, the client uses `HttpArtifactRepository` (https://github.com/mlflow/mlflow/blob/3c434246f73b41835194f3e6aa1f9be8488bcd56/mlflow/store/artifact/http_artifact_repo.py#L11) to upload and download files to / from the remote host where the MLflow Tracking server is running.

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:836]: great call-out. Modified it here and a few other places.

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:1034]: changed 17 references.

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:1048]: updated both of these references

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:1053]: rephrased the entire paragraph.

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:843]: definitely. removed them.

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:838]: Added a few listed examples - please check to make sure I'm capturing what you're looking for here or if I completely missed the plot.

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:189]: updated!

File: docs/source/tracking.rst
Comment (BenWilson2) [reply docs/source/tracking.rst:206]: fixed the wording here.

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:217]:
```suggestion
    Administrators who are enabling this feature should ensure that the access level granted to the Tracking Server for artifact operations meets all
    security requirements prior to enabling the Tracking Server to operate in a proxied file handling role.
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:226]: "bastion host" is somewhat bespoke terminology. Can we simplify the phrasing here? 

```suggestion
MLflow's Tracking Server can be used in an exclusive proxied artifact handling role. Specifying the
``--artifacts-only`` flag restricts an MLflow server instance to only serve artifact-related API requests by
proxying to an underlying object store.
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:234]: I think this is already stated in the `note` immediately above & can be removed.

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:758]:
```suggestion
An MLflow Tracking server can also be run as a proxied artifact handler. An example configuration for the ``mlflow server`` in this mode is:
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:838]:
```suggestion
Starting a server with the ``--serve-artifacts`` flag enables proxied access for artifacts.
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:851]: Can we clarify that, when the host is absent from the `mlflow-artifacts` URI, the client assumes that the host is the same as the host component of the tracking URI?

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:1047]:
```suggestion
To use an instance of the MLflow Tracking server *exclusively* for artifact operations ( :ref:`scenario_6` ),
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:1049]:
```suggestion
for disabling all other functionality of the Tracking server.
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:1062]:
```suggestion
Using an MLflow server configured in ``--artifacts-only`` mode for any tasks aside from those concerned with artifact
```

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:1080]: I think the biggest motivator is decoupling the compute / infrastructure used for serving these different types of requests, since each type of request (artifacts vs metadata) has different performance requirements and usage characteristics.

**Review by dbczumar — APPROVED** [review]
@BenWilson2 Left a few more comments. LGTM once they're addressed - feel free to merge after that. Thanks so much!

**Pull request merged**
